baseServer.go

```diff
@@ -0,0 +1,208 @@
+package basePlatformSOMAS
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+type BaseServer[T IAgent[T]] struct {
+	// map of agentid -> agent struct
+	agentMap map[uuid.UUID]T
+	// map of agentid -> empty struct so that agents cannot access each others agent structs
+	agentIdSet map[uuid.UUID]struct{}
+	// waitgroup for message synchronisation
+	listeningWaitGroup *sync.WaitGroup
+	// map of uuid ->struct{}{} which stores the ids of agents which have stopped messaging
+	agentStoppedTalkingMap map[uuid.UUID]struct{}
+	messagingFinished      chan struct{}
+	// duration after which messaging phase forcefully ends during rounds
+	turnTimeout           time.Duration
+	messageHandlerTimeout time.Duration
+	// run single round
+	roundRunner RoundRunner
+	//iterations
+	iterations int
+	turns      int
+}
+
+func (server *BaseServer[T]) HandleStartOfTurn(iter, round int) {
+	fmt.Printf("Iteration %d, Round %d starting...\n", iter, round)
+
+}
+
+func (serv *BaseServer[T]) endAgentListeningSession() (string, bool) {
+	
+	ctx, cancel := context.WithTimeout(context.Background(), serv.turnTimeout)
+	defer cancel()
+	select {
+	case <-serv.messagingFinished:
+		return "finished messaging", true
+
+	case <-ctx.Done():
+		return "timed out before all messages sent", false
+	}
+}
+
+func (server *BaseServer[T]) HandleEndOfTurn(iter, round int) {
+	server.endAgentListeningSession()
+	fmt.Printf("Iteration %d, Round %d finished.\n", iter, round)
+}
+
+func (server *BaseServer[T]) RunAgentLoop() {}
+
+func (server *BaseServer[T]) messageHandlerLimiter(msg IMessage[T], receiver uuid.UUID) {
+	ctx, cancel := context.WithTimeout(context.Background(), server.messageHandlerTimeout)
+	defer cancel()
+	go msg.InvokeMessageHandler(server.agentMap[receiver])
+	<-ctx.Done()
+
+}
+
+func (server *BaseServer[T]) SendMessage(msg IMessage[T], receivers []uuid.UUID) {
+	for _, receiver := range receivers {
+		go server.messageHandlerLimiter(msg, receiver)
+	}
+
+}
+
+func (serv *BaseServer[T]) AddAgent(agent T) {
+	serv.agentMap[agent.GetID()] = agent
+	serv.agentIdSet[agent.GetID()] = struct{}{}
+}
+
+func (serv *BaseServer[T]) ViewAgentIdSet() map[uuid.UUID]struct{} {
+	return serv.agentIdSet
+}
+
+func (serv *BaseServer[T]) AccessAgentByID(id uuid.UUID) T {
+	return serv.agentMap[id]
+}
+
+func (serv *BaseServer[T]) Initialise() {}
+
+func (serv *BaseServer[T]) Start() {
+	serv.checkHandler()
+	turns := serv.turns
+	iterations := serv.iterations
+	for i := 0; i < iterations; i++ {
+		for j := 0; j < turns; j++ {
+			serv.HandleStartOfTurn(i+1, j+1)
+			serv.roundRunner.RunTurn()
+			serv.HandleEndOfTurn(i+1, j+1)
+		}
+	}
+}
+
+func (serv *BaseServer[T]) GetAgentMap() map[uuid.UUID]T {
+	return serv.agentMap
+}
+
+func (serv *BaseServer[T]) agentStoppedTalking(id uuid.UUID) {
+	fmt.Println("sending stop talking request,id:", id)
+	serv.agentStoppedTalkingMap[id] = struct{}{}
```

Comment:
can we put a mutex around this? I think it's much safer

Reply:
we 'technically' have a race condition here, where one agent thread might update the map at the same time we read from it. Adding a simple mutex is much cleaner